src/common/fs_manager.cpp

```diff
@@ -344,5 +344,30 @@ void fs_manager::update_disk_stat()
     _counter_min_available_ratio->set(_min_available_ratio);
     _counter_max_available_ratio->set(_max_available_ratio);
 }
+
+void fs_manager::add_new_dir_node(const std::string &data_dir, const std::string &tag)
+{
+    zauto_write_lock l(_lock);
+    std::string norm_path;
+    utils::filesystem::get_normalized_path(data_dir, norm_path);
+    dir_node *n = new dir_node(tag, norm_path);
+    _dir_nodes.emplace_back(n);
+    _available_data_dirs.emplace_back(data_dir);
+    ddebug_f("{}: mark data dir({}) as tag({})", dsn_primary_address().to_string(), norm_path, tag);
+}
+
+bool fs_manager::is_dir_node_available(const std::string &data_dir, const std::string &tag) const
+{
+    zauto_read_lock l(_lock);
+    for (const auto &dir_node : _dir_nodes) {
+        std::string norm_path;
+        utils::filesystem::get_normalized_path(data_dir, norm_path);
+        if (dir_node->full_dir == norm_path || dir_node->tag == tag) {
```

Comment:
why not &&

Reply:
The new dir_node should not has same same path or tag with the current nodes. For example, if the new node dir path=new_path, tag=tag7, if the current dir node(path = new_path, tag=tag6), the new dir node should not be added, and if the current dir node(path = old_path, tag=tag7), the new dir node also should not be added.

Reply:
if (dir_node->full_dir == norm_path || dir_node->tag == tag) {
return false
}
return true

return false only has same case?

Reply:
Oh, I know your misunderstading. This function name is `is_dir_node_available` to check if dir exist, if it exists, return true, otherwise return false. The caller `on_add_new_disk` will check its return value, if it is true, reject adding this disk, you can reference L1158-L1163 of replica_stub.cpp added in this pr.